src/main/java/com/google/sps/servlets/FavoriteRecipeServlet.java

```diff
@@ -0,0 +1,151 @@
+        PreparedQuery results = datastore.prepare(query);
+      
+        Entity entity = results.asSingleEntity();
+        if (entity == null) {
+          responseMap.put("isFavorite", false);
+        } else {
+          responseMap.put("isFavorite", true);
+          long favoriteId = entity.getKey().getId();
+          responseMap.put("favoriteId", favoriteId);
+        }
+      } else {
+        // give back list of favorited recipes
+        query.setFilter(new Query.FilterPredicate("userId", Query.FilterOperator.EQUAL, userService.getCurrentUser().getUserId()));
+        PreparedQuery results = datastore.prepare(query);
+
+        ArrayList<Long> recipeList = new ArrayList<>();
+        for (Entity entity : results.asIterable()) {
+          long entityRecipeId = (long) entity.getProperty("recipeId");
+          recipeList.add(entityRecipeId);
+        }
+        responseMap.put("recipeList", recipeList);
+      }
+
+    } else {
+      responseMap.put("error", AUTHORIZATION_ERROR); 
+    }
+    
+    json = gson.toJson(responseMap);
+    response.setContentType("application/json");
+    response.getWriter().println(json);
+  }
+
+  /** Add/Delete a FavoriteRecipe to datastore which represents a user's favorited recipe */
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    UserService userService = UserServiceFactory.getUserService();
+    JSONObject responseMap = new JSONObject();
+
+    if (!userService.isUserLoggedIn()) {
+      responseMap.put("error", AUTHORIZATION_ERROR);
+    } else {
+      String favoriteIdString = request.getParameter("favorite-id");
+      if (favoriteIdString != null && favoriteIdString != "0") {
+        // delete entity
+        long id = Long.parseLong(favoriteIdString);
+
+        Key tagEntityKey = KeyFactory.createKey("FavoriteRecipe", id);
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        datastore.delete(tagEntityKey);
+      } else {
+        // create entity
+        Entity entity = new Entity("FavoriteRecipe");
+      
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        String userId = userService.getCurrentUser().getUserId();
+        entity.setProperty("userId", userId);
+      
+        Long recipeId = strToLong(request.getParameter("recipe-id"));
+        entity.setProperty("recipeId", recipeId);
+      
+        if (recipeId != null) {
+          datastore.put(entity);
+
+          long favoriteId = entity.getKey().getId();
+          responseMap.put("favoriteId", favoriteId);
+        }
+      }
+    }
+
+    Gson gson = new Gson();   
+    String json = gson.toJson(responseMap);
+    response.setContentType("application/json");
+    response.getWriter().println(json);
+  }
+
+  public Long strToLong(String str) {
+    if (str == null) {
+      return null;
+    } else {
+      return Long.parseLong(str);
+    }
+  }
```

Comment:
Consider moving this into a util class, since it's used at several places in different servlets.